src/primitive_core.cc

```diff
@@ -65,14 +65,14 @@ PRIMITIVE(write_string_on_stdout) {
   ARGS(cstring, message, bool, add_newline);
   fprintf(stdout, "%s%s", message, add_newline ? "\n" : "");
   fflush(stdout);
-  return _raw_message;
+  return process->program()->null_object();
```

Comment:
Could be worth doing these changes independently of the type propagation, but I think it's ok.